// @ceramicnetwork/rpc-window/dist/rpc-window.cjs.development.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var rpcPostmessage = require('@ceramicnetwork/rpc-postmessage');
var transportPostmessage = require('@ceramicnetwork/transport-postmessage');

var DEFAULT_OPTIONS = {
  onInvalidInput: function onInvalidInput(_input, _error) {// Silence warnings of invalid messages, such as message events sent by third-parties
  },
  // Client requests will be sent without origin restriction
  postMessageArguments: ['*']
};
function createClient(namespace, target, options) {
  if (target === void 0) {
    target = window;
  }

  if (options === void 0) {
    options = DEFAULT_OPTIONS;
  }

  var transport = transportPostmessage.createPostMessageTransport(window, target, options);
  return rpcPostmessage.createNamespaceClient(transport, namespace, options);
}
function createServer(namespace, methods, target) {
  if (target === void 0) {
    target = window;
  }

  return rpcPostmessage.createNamespaceServer({
    methods: methods,
    namespace: namespace,
    target: target
  });
}

exports.createClient = createClient;
exports.createServer = createServer;
