// ipfs-repo/lib/stores/locks.js
'use strict';

exports = module.exports;

exports.setUp = function (basePath, blobStore) {
  var store = blobStore(basePath);
  var lockFile = 'repo.lock';

  return {
    lock: function lock(cb) {
      function createLock() {
        store.createWriteStream(lockFile).once('finish', function () {
          cb();
        }).end();
      }

      function doesExist(err, exists) {
        if (err) {
          return cb(err);
        }
        if (exists) {
          // default 100ms
          setTimeout(function () {
            store.exists(lockFile, doesExist);
          }, 100);
          return;
        }
        createLock();
      }

      store.exists(lockFile, doesExist);
    },
    unlock: function unlock(cb) {
      store.remove(lockFile, function (err) {
        if (err) {
          return cb(err);
        }
        store.exists(lockFile, function (err, exists) {
          if (err) {
            return cb(err);
          }

          store.exists(lockFile, function (err, exists) {
            if (err) {
              return cb(err);
            }
            cb();
          });
        });
      });
    }
  };
};
