// @ceramicnetwork/blockchain-utils-linking/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SolanaAuthProvider = exports.TezosAuthProvider = exports.NearAuthProvider = exports.CosmosAuthProvider = exports.PolkadotAuthProvider = exports.EosioAuthProvider = exports.FilecoinAuthProvider = exports.AvalancheAuthProvider = exports.EthereumAuthProvider = exports.solana = exports.tezos = exports.cosmos = exports.eosio = exports.polkadot = exports.filecoin = exports.ethereum = void 0;
__exportStar(require("./auth-provider"), exports);
exports.ethereum = __importStar(require("./ethereum"));
__exportStar(require("./util"), exports);
exports.filecoin = __importStar(require("./filecoin"));
exports.polkadot = __importStar(require("./polkadot"));
exports.eosio = __importStar(require("./eosio"));
exports.cosmos = __importStar(require("./cosmos"));
exports.tezos = __importStar(require("./tezos"));
exports.solana = __importStar(require("./solana"));
var ethereum_1 = require("./ethereum");
Object.defineProperty(exports, "EthereumAuthProvider", { enumerable: true, get: function () { return ethereum_1.EthereumAuthProvider; } });
var ethereum_2 = require("./ethereum");
Object.defineProperty(exports, "AvalancheAuthProvider", { enumerable: true, get: function () { return ethereum_2.EthereumAuthProvider; } });
var filecoin_1 = require("./filecoin");
Object.defineProperty(exports, "FilecoinAuthProvider", { enumerable: true, get: function () { return filecoin_1.FilecoinAuthProvider; } });
var eosio_1 = require("./eosio");
Object.defineProperty(exports, "EosioAuthProvider", { enumerable: true, get: function () { return eosio_1.EosioAuthProvider; } });
var polkadot_1 = require("./polkadot");
Object.defineProperty(exports, "PolkadotAuthProvider", { enumerable: true, get: function () { return polkadot_1.PolkadotAuthProvider; } });
var cosmos_1 = require("./cosmos");
Object.defineProperty(exports, "CosmosAuthProvider", { enumerable: true, get: function () { return cosmos_1.CosmosAuthProvider; } });
var near_1 = require("./near");
Object.defineProperty(exports, "NearAuthProvider", { enumerable: true, get: function () { return near_1.NearAuthProvider; } });
var tezos_1 = require("./tezos");
Object.defineProperty(exports, "TezosAuthProvider", { enumerable: true, get: function () { return tezos_1.TezosAuthProvider; } });
var solana_1 = require("./solana");
Object.defineProperty(exports, "SolanaAuthProvider", { enumerable: true, get: function () { return solana_1.SolanaAuthProvider; } });
